Tests/PackageCollectionsTests/JSONPackageCollectionProviderTests.swift

```diff
@@ -249,7 +249,7 @@ class JSONPackageCollectionProviderTests: XCTestCase {
         let handler: HTTPClient.Handler = { request, _, completion in
             XCTAssertEqual(request.url, url, "url should match")
             XCTAssertEqual(request.method, .head, "method should match")
-            completion(.success(.init(statusCode: statusCode)))
+            completion(.failure(HTTPClientError.badResponseStatusCode(statusCode)))
```

Comment:
these changes seems like unintended consequences. why would we need to do this?

Reply:
The status quo can be modeled with this diagram, where `HTTPClient` delegates to `URLSessionHTTPClient` enhancing it with Dispatch helpers (semaphores, scheduling on specific `OperationQueue` etc), and also adds HTTP status code validation and a few more smarts.

```
            ┏━━━━━━━━━━━━━━━━━━━━━━━┓           
            ┃      HTTPClient       ┃           
┏━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━┓
┃                                              ┃
┃       Dispatch multi-threading helpers       ┃
┃                                              ┃
┃─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┃
┃                                              ┃
┃      Circuit breaker and retry strategies    ┃
┃                                              ┃
┃─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┃
┃                                              ┃
┃         HTTP status code validation          ┃
┃                                              ┃
┗━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┛
                        │                       
                        ▼                       
           ┏━━━━━━━━━━━━━━━━━━━━━━━━┓           
           ┃  URLSessionHTTPClient  ┃           
┏━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━┓
┃                                              ┃
┃              URLSession wrapper              ┃
┃                                              ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

Dispatch helpers become irrelevant with Swift Concurrency though, and I don't think it won't be easy to make them optional in `HTTPClient` without a heavy refactoring. What I'm proposing in this PR is adding an `async` wrapper on `URLSessionHTTPClient` level, where Dispatch assumptions don't leak outside to the API surface. This way `async`-friendly code can start using `URLSessionHTTPClient` directly instead of "proxying" through `HTTPClient`, as I'm not sure if indirection introduced originally with `HTTPClient` is helpful outside of Dispatch context. 

HTTP status code validation is quite useful though, so I'm moving it to that lower level too. What we currently have in this PR is shown in this diagram:

```
            ┏━━━━━━━━━━━━━━━━━━━━━━━┓           
            ┃      HTTPClient       ┃           
┏━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━┓
┃                                              ┃
┃       Dispatch multi-threading helpers       ┃
┃                                              ┃
┃─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┃
┃                                              ┃
┃      Circuit breaker and retry strategies    ┃
┃                                              ┃
┗━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┛
                        │                       
                        ▼                       
           ┏━━━━━━━━━━━━━━━━━━━━━━━━┓           
           ┃  URLSessionHTTPClient  ┃           
┏━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━┓
┃                                              ┃
┃       async wrapper for `func execute`       ┃
┃                                              ┃
┃─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┃
┃                                              ┃
┃         HTTP status code validation          ┃
┃                                              ┃
┃─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┃
┃                                              ┃
┃              URLSession wrapper              ┃
┃                                              ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

Since some tests configure `HTTPClient` to delegate not to `URLSessionHTTPClient`, but to a mock handler instead, they lose HTTP status code validation. It happens only in those few tests specifically. To my knowledge, in non-test targets of SwiftPM we only use `HTTPClient` with `URLSessionHTTPClient` as a handler, so that behavior stays intact.

We could argue that circuit breaker and retry strategies should be moved away from `HTTPClient` too, but that would need an even bigger refactoring. Would you like that to be done in this PR too, or do you want to do the opposite and split this one into smaller PR? WDYT of the overall approach?